### PR TITLE
New azure_rm_privatednszone module with tests

### LINKED ISSUE #87: azure_rm_dnszone fails when using private DNS zones
##### SUMMARY

Trying to create a private DNS Zone:

$ ansible -i inventory localhost -m azure_rm_dnszone -a "resource_group=rg name=private.example.com state=present type=private"
[WARNING]: Azure API profile latest does not define an entry for DnsManagementClient
localhost | FAILED! => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "changed": false,
    "msg": "Error creating or updating zone rhocp.us - Operation failed with status: 'Bad Request'. Details: The feature you are trying to use is not available."
}

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

azure_rm_dnszone

##### ANSIBLE VERSION

```paste below
$ ansible --version
ansible 2.9.6
  config file = /home/ocp/azure/ansible/ansible.cfg
  configured module search path = ['/home/ocp/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.6 (default, Jan 30 2020, 09:44:41) [GCC 9.2.1 20190827 (Red Hat 9.2.1-1)]
```

##### CONFIGURATION

```paste below
$ ansible-config dump --only-changed
COMMAND_WARNINGS(/home/ocp/azure/ansible/ansible.cfg) = False
DEFAULT_FILTER_PLUGIN_PATH(/home/ocp/azure/ansible/ansible.cfg) = ['/home/ocp/azure/ansible/filter_plugins']
DEFAULT_HOST_LIST(/home/ocp/azure/ansible/ansible.cfg) = ['/home/ocp/azure/ansible/inventory']
DEPRECATION_WARNINGS(/home/ocp/azure/ansible/ansible.cfg) = False
HOST_KEY_CHECKING(/home/ocp/azure/ansible/ansible.cfg) = False
INTERPRETER_PYTHON(/home/ocp/azure/ansible/ansible.cfg) = auto
```

##### OS / ENVIRONMENT

Fedora 31

##### STEPS TO REPRODUCE

Run the command in the summary 

```yaml
$ ansible -i inventory localhost -m azure_rm_dnszone -a "resource_group=rg name=private.example.com state=present type=private"
```

##### EXPECTED RESULTS

Zone created or data for existing zone returned.

##### ACTUAL RESULTS

Error: The feature you are trying to use is not available.

```paste below
localhost | FAILED! => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "changed": false,
    "msg": "Error creating or updating zone rhocp.us - Operation failed with status: 'Bad Request'. Details: The feature you are trying to use is not available."
}
```

### PR DESCRIPTION
**Pull request by imjoseangel**
##### SUMMARY

New azure_rm_privatednszone module.
Fixes #87 

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

azure_rm_privatednszone

##### ADDITIONAL INFORMATION

Creates Private DNS Zone in Azure with the brand new privatedns Python module from Azure SDK

```yaml
    - name: Create a private DNS zone (check mode)
      azure_rm_privatednszone:
        resource_group: "{{ resource_group }}"
        name: "{{ domain_name }}.com"
      register: results
      check_mode: true
```

**Note:** The Azure SDK is still on development. We will add new functionality to the module as soon as it is coming. A new azure_rm_privatednszone_info module will be released as soon as this is approved.

### CODE REVIEW

**Comment by Fred-sun** [comment]
@imjoseangel Thank you for your contribution to ansible-collection, can you add azure_rm_privatednszone_info module to privatenszone? Thank you very much!

**Comment by imjoseangel** [comment]
Hi @Fred-sun Do you mean both modules in the same PR?

Thank you!

**Comment by Fred-sun** [comment]
@imjoseangel Yes. Can be in the same PR, because they share "integration/targets/azure_rm_privatednszone/tasks/main.yml".

File: plugins/modules/azure_rm_privatednszone.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+short_description: Manage Azure private DNS zones
+
+description:
+    - Creates and deletes Azure private DNS zones.
+
+options:
+    resource_group:
+        description:
+            - name of resource group.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_privatednszone.py]:
```suggestion
            - Name of resource group.
```

Comment (imjoseangel) [reply plugins/modules/azure_rm_privatednszone.py]: Changed. Thanks!

**Comment by Fred-sun** [comment]
@haiyuazhang ready_for_review

File: plugins/modules/azure_rm_privatednszone_info.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+            - Limit results by resource group. Required when filtering by name.
+    name:
+        description:
+            - Only show results for a specific zone.
+    tags:
+        description:
+            - Limit results by providing a list of tags. Format tags as 'key' or 'key:value'.
+
+extends_documentation_fragment:
+    - azure
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_privatednszone_info.py]:
```suggestion
    - azure.azcollection.azure
```

File: plugins/modules/azure_rm_privatednszone_info.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    - azure_tags
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_privatednszone_info.py]:
```suggestion
    - azure.azcollection.azure_tags
```

File: plugins/modules/azure_rm_privatednszone_info.py (hunk 2 above)
Comment (imjoseangel) [reply plugins/modules/azure_rm_privatednszone_info.py]: Changed! Thanks!

File: plugins/modules/azure_rm_privatednszone_info.py (hunk 3 above)
Comment (imjoseangel) [reply plugins/modules/azure_rm_privatednszone_info.py]: Changed Thanks!

**Comment by imjoseangel** [comment]
@Fred-sun Is this module going to be merged or there is anything else to implement?

Thank you

**Comment by Fred-sun** [comment]
@imjoseangel I'm sorry that this PR does not Merge yet, because there are many things happening recently, developers are busy with the launch of other modules, and they ignore this PR. I will push PR Merge as soon as possible. Thank you very much!

**Comment by Fred-sun** [comment]
@imjoseangel The document format needs to be changed. Can you Fix the document changes according to the following prompt? Thank you very much!
```e.g.
ERROR: plugins/modules/azure_rm_privatednszone.py:1:1: expected module shebang "b'#!/usr/bin/python'" but found: b'#!/usr/bin/env python'
ERROR: plugins/modules/azure_rm_privatednszone_info.py:1:1: expected module shebang "b'#!/usr/bin/python'" but found: b'#!/usr/bin/env python'
ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: missing-python-interpreter: Interpreter line is not "#!/usr/bin/python"
ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: parameter-type-not-in-doc: Argument 'name' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: parameter-type-not-in-doc: Argument 'resource_group' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: parameter-type-not-in-doc: Argument 'state' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: doc-type-does-not-match-spec: Argument 'tags' in argument_spec defines type as 'list' but documentation defines type as 'dict'
ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: missing-python-interpreter: Interpreter line is not "#!/usr/bin/python"
ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: parameter-type-not-in-doc: Argument 'name' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: parameter-type-not-in-doc: Argument 'resource_group' in argument_spec defines type as 'str' but documentation doesn't define type

**Comment by imjoseangel** [comment]
@Fred-sun Changed and tested with `ansible-test sanity -v --docker --python 2.7`

Thank you!

**Comment by Fred-sun** [comment]
@imjoseangel There are some other document errors that need your help to fix, please refer to the error tips below. Thank you very much!
Link: https://dev.azure.com/mseng/VSOSS/_build/results?buildId=12471813&view=logs&j=76c949d7-ae3c-5dab-f3c6-f88fb1d3b538&t=8d517798-7997-54bb-f3b0-245ccd1cc734
```e.g.

ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: deprecation-either-date-or-version: DOCUMENTATION.options.adfs_authority_url: version_added ('2.6') is not a valid collection version (see specification at https://semver.org/): invalid semantic version '2.6' for dictionary value @ data['options']['adfs_authority_url']. Got AnsibleMapping([('description', ['Azure AD authority url. Use when authenticating with Username/password, and has your own ADFS authority.']), ('type', 'str'), ('version_added', '2.6'), ('version_added_collection', 'azure.azcollection')])

ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: deprecation-either-date-or-version: DOCUMENTATION.options.api_profile: version_added ('2.5') is not a valid collection version (see specification at https://semver.org/): invalid semantic version '2.5' for dictionary value @ data['options']['api_profile']. Got AnsibleMapping([('description', ['Selects an API profile to use when communicating with Azure services. Default value of C(latest) is appropriate for public clouds; future values will allow use with Azure Stack.']), ('type', 'str'), ('default', 'latest'), ('version_added', '2.5'), ('version_added_collection', 'azure.azcollection')])

ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: deprecation-either-date-or-version: DOCUMENTATION.options.auth_source: version_added ('2.5') is not a valid collection version (see specification at https://semver.org/): invalid semantic version '2.5' for dictionary value @ data['options']['auth_source']. Got AnsibleMapping([('description', ['Controls the source of the credentials to use for authentication.', 'If not specified, ANSIBLE_AZURE_AUTH_SOURCE environment variable will be used and default to C(auto) if variable is not defined.', 'C(auto) will follow the default precedence of module parameters -> environment variables -> default profile in credential file C(~/.azure/credentials).', 'When set to C(cli), the credentials will be sources from the default Azure CLI profile.', 'Can also be set ...

ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: deprecation-either-date-or-version: DOCUMENTATION.options.cert_validation_mode: version_added ('2.5') is not a valid collection version (see specification at https://semver.org/): invalid semantic version '2.5' for dictionary value @ data['options']['cert_validation_mode']. Got AnsibleMapping([('description', ['Controls the certificate validation behavior for Azure endpoints. By default, all modules will validate the server certificate, but when an HTTPS proxy is in use, or against Azure Stack, it may be necessary to disable this behavior by passing C(ignore). Can also be set via credential file profile or the C(AZURE_CERT_VALIDATION) environment variable.']), ('type', 'str'), ('choices', ['ignore', 'validate']), ('version_added', '2.5'), ('version_added_collection'...

ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: deprecation-either-date-or-version: DOCUMENTATION.options.cloud_environment: version_added ('2.4') is not a valid collection version (see specification at https://semver.org/): invalid semantic version '2.4' for dictionary value @ data['options']['cloud_environment']. Got AnsibleMapping([('description', ['For cloud environments other than the US public cloud, the environment name (as defined by Azure Python SDK, eg, C(AzureChinaCloud), C(AzureUSGovernment)), or a metadata discovery endpoint URL (required for Azure Stack). Can also be set via credential file profile or the C(AZURE_CLOUD_ENVIRONMENT) environment variable.']), ('type', 'str'), ('default', 'AzureCloud'), ('version_added', '2.4'), ('version_added_collection', 'azure.azcollection')])

ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: required_if-requirements-unknown: required_if contains terms in requirements which are not part of argument_spec: log_path


ERROR: plugins/modules/azure_rm_privatednszone.py:0:0: required_if-unknown-key: required_if must have its key log_mode in argument_spec

ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: deprecation-either-date-or-version: DOCUMENTATION.options.adfs_authority_url: version_added ('2.6') is not a valid collection version (see specification at https://semver.org/): invalid semantic version '2.6' for dictionary value @ data['options']['adfs_authority_url']. Got AnsibleMapping([('description', ['Azure AD authority url. Use when authenticating with Username/password, and has your own ADFS authority.']), ('type', 'str'), ('version_added', '2.6'), ('version_added_collection', 'azure.azcollection')])

ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: deprecation-either-date-or-version: DOCUMENTATION.options.api_profile: version_added ('2.5') is not a valid collection version (see specification at https://semver.org/): invalid semantic version '2.5' for dictionary value @ data['options']['api_profile']. Got AnsibleMapping([('description', ['Selects an API profile to use when communicating with Azure services. Default value of C(latest) is appropriate for public clouds; future values will allow use with Azure Stack.']), ('type', 'str'), ('default', 'latest'), ('version_added', '2.5'), ('version_added_collection', 'azure.azcollection')])

ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: deprecation-either-date-or-version: DOCUMENTATION.options.auth_source: version_added ('2.5') is not a valid collection version (see specification at https://semver.org/): invalid semantic version '2.5' for dictionary value @ data['options']['auth_source']. Got AnsibleMapping([('description', ['Controls the source of the credentials to use for authentication.', 'If not specified, ANSIBLE_AZURE_AUTH_SOURCE environment variable will be used and default to C(auto) if variable is not defined.', 'C(auto) will follow the default precedence of module parameters -> environment variables -> default profile in credential file C(~/.azure/credentials).', 'When set to C(cli), the credentials will be sources from the default Azure CLI profile.', 'Can also be set ...

ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: deprecation-either-date-or-version: DOCUMENTATION.options.cert_validation_mode: version_added ('2.5') is not a valid collection version (see specification at https://semver.org/): invalid semantic version '2.5' for dictionary value @ data['options']['cert_validation_mode']. Got AnsibleMapping([('description', ['Controls the certificate validation behavior for Azure endpoints. By default, all modules will validate the server certificate, but when an HTTPS proxy is in use, or against Azure Stack, it may be necessary to disable this behavior by passing C(ignore). Can also be set via credential file profile or the C(AZURE_CERT_VALIDATION) environment variable.']), ('type', 'str'), ('choices', ['ignore', 'validate']), ('version_added', '2.5'), ('version_added_collection'...

ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: deprecation-either-date-or-version: DOCUMENTATION.options.cloud_environment: version_added ('2.4') is not a valid collection version (see specification at https://semver.org/): invalid semantic version '2.4' for dictionary value @ data['options']['cloud_environment']. Got AnsibleMapping([('description', ['For cloud environments other than the US public cloud, the environment name (as defined by Azure Python SDK, eg, C(AzureChinaCloud), C(AzureUSGovernment)), or a metadata discovery endpoint URL (required for Azure Stack). Can also be set via credential file profile or the C(AZURE_CLOUD_ENVIRONMENT) environment variable.']), ('type', 'str'), ('default', 'AzureCloud'), ('version_added', '2.4'), ('version_added_collection', 'azure.azcollection')])

ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: parameter-list-no-elements: Argument 'tags' in argument_spec defines type as list but elements is not defined

ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: required_if-requirements-unknown: required_if contains terms in requirements which are not part of argument_spec: log_path

ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: required_if-unknown-key: required_if must have its key log_mode in argument_spec

ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: return-syntax-error: RETURN.privatednszones.contains.id.type: required key not provided @ data['privatednszones']['contains']['id']['type']. Got None

ERROR: plugins/modules/azure_rm_privatednszone_info.py:0:0: return-syntax-error: RETURN.privatednszones.contains.name.type: required key not provided @ data['privatednszones']['contains']['name']['type']. Got None

See documentation for help: https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/validate-modules.html

File: plugins/modules/azure_rm_privatednszone_info.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+        super(AzurePrivateRMDNSZoneInfo, self).__init__(self.module_arg_spec)
+
+    def exec_module(self, **kwargs):
+
+        is_old_facts = self.module._name == 'azure_rm_privatednszone_facts'
+        if is_old_facts:
+            self.module.deprecate(
+                "The 'azure_rm_privatednszone_facts' module has been renamed to 'azure_rm_privatednszone_info'",
+                version='2.13')
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_privatednszone_info.py]:
```suggestion
                version=(2.9, ))
```

**Comment by imjoseangel** [comment]
Thank you @Fred-sun Is it possible checking those flags locally or getting access to the online resource? 

Regards

**Comment by Fred-sun** [comment]
> Thank you @Fred-sun Is it possible checking those flags locally or getting access to the online resource?
> Regards

@imjoseangel Sorry, I don't know, because I'm always debugging these errors online. Thank you very much!

**Comment by Fred-sun** [comment]
@imjoseangel I made some changes to the devel version test fail. Can you merge the changes I submitted into your PR? Thank you!
Link: https://github.com/imjoseangel/azure/pull/1

**Pull request merged**